src/libproc/proc_pid.rs

```diff
@@ -503,3 +503,132 @@ fn name_test_init_pid() {
+/// The length of return value: `Vec<T::Item>` may be less than `max_len`.
+///
+/// # Examples
+///
+/// ```
+/// use std::io::Write;
+/// use libproc::libproc::proc_pid::{listpidinfo, pidinfo, ListFDs, ListThreads, TaskAllInfo};
+///
+/// fn listthreads_test() {
+///     use std::process;
+///     let pid = process::id() as i32;
+///
+///     match pidinfo::<TaskAllInfo>(pid, 0) {
+///         Ok(info) => {
+///             match listpidinfo::<ListThreads>(pid, info.ptinfo.pti_threadnum as usize) {
+///                 Ok(threads) => assert!(threads.len()>0),
+///                 Err(err) => assert!(false, "Error retrieving process info: {}", err)
+///             }
+///             match listpidinfo::<ListFDs>(pid, info.pbsd.pbi_nfiles as usize) {
+///                 Ok(fds) => assert!(fds.len()>0),
+///                 Err(err) => assert!(false, "Error retrieving process info: {}", err)
+///             }
+///         },
+///         Err(err) => assert!(false, "Error retrieving process info: {}", err)
+///     };
+/// }
+/// ```
+pub fn listpidinfo<T: ListPIDInfo>(pid : i32, max_len: usize) -> Result<Vec<T::Item>, String> {
+    let flavor = T::flavor() as i32;
+    let buffer_size = mem::size_of::<T::Item>() as i32 * max_len as i32;
+    let mut buffer = Vec::<T::Item>::with_capacity(max_len);
+    let buffer_ptr = unsafe {
+        buffer.set_len(max_len);
+        buffer.as_mut_ptr() as *mut c_void
+    };
+
+    let ret: i32;
+
+    unsafe {
+        ret = proc_pidinfo( pid, flavor, 0, buffer_ptr, buffer_size);
+    };
+
+    if ret <= 0 {
+        Err(get_errno_with_message(ret))
+    } else {
+        let actual_len = ret as usize / mem::size_of::<T::Item>();
+        buffer.truncate(actual_len);
+        Ok(buffer)
+    }
+}
+
+#[test]
+fn listpidinfo_test() {
+    use std::process;
+    let pid = process::id() as i32;
+
+    match pidinfo::<TaskAllInfo>(pid, 0) {
+        Ok(info) => {
+            match listpidinfo::<ListThreads>(pid, info.ptinfo.pti_threadnum as usize) {
+                Ok(threads) => assert!(threads.len()>0),
+                Err(err) => assert!(false, "Error retrieving process info: {}", err)
+            }
+            match listpidinfo::<ListFDs>(pid, info.pbsd.pbi_nfiles as usize) {
+                Ok(fds) => assert!(fds.len()>0),
+                Err(err) => assert!(false, "Error retrieving process info: {}", err)
+            }
+        },
+        Err(err) => assert!(false, "Error retrieving process info: {}", err)
+    };
+}
+
+pub struct ListThreads;
+
+impl ListPIDInfo for ListThreads {
+    type Item = uint64_t;
+    fn flavor() -> PidInfoFlavor { PidInfoFlavor::ListThreads }
+}
+
+pub struct ListFDs;
+
+impl ListPIDInfo for ListFDs {
+    type Item = ProcFDInfo;
+    fn flavor() -> PidInfoFlavor { PidInfoFlavor::ListFDs }
+}
+
+#[repr(C)]
+pub struct ProcFDInfo {
+    pub proc_fd: int32_t,
+    pub proc_fdtype: uint32_t,
+}
+
+#[derive(Copy, Clone, Debug)]
```

Comment:
WDYT - are all these acronyms already familiar to users of the lib (from apple docs etc)?
Or should we rename, or add comments to explain them?

Reply:
This is from https://opensource.apple.com/source/xnu/xnu-1504.7.4/bsd/sys/proc_info.h
```
/* defns of process file desc type */
#define PROX_FDTYPE_ATALK	0
#define PROX_FDTYPE_VNODE	1
#define PROX_FDTYPE_SOCKET	2
#define PROX_FDTYPE_PSHM	3
#define PROX_FDTYPE_PSEM	4
#define PROX_FDTYPE_KQUEUE	5
#define PROX_FDTYPE_PIPE	6
#define PROX_FDTYPE_FSEVENTS	7
```
But I don't know the meaning of the acronyms...
Do you know any documentation?

Reply:
I added doc comment.
Some acronyms ( vnode/kqueue/FSEvents ) seems to be generally used in macOS or BSD, so I kept these.